File: Packages/mcp-unity/Editor/Tools/PrefabOpenTool.cs
using System;
using UnityEngine;
using UnityEditor;
using Newtonsoft.Json.Linq;
using McpUnity.Unity;
using McpUnity.Utils;

namespace McpUnity.Tools
{
    /// <summary>
    /// Tool for opening a prefab asset in Prefab Mode.
    /// </summary>
    public class PrefabOpenTool : McpToolBase
    {
        public PrefabOpenTool()
        {
            Name = "prefab_open";
            Description = "Opens a prefab asset in Prefab Mode.";
        }

        public override JObject Execute(JObject parameters)
        {
            string prefabPath = parameters?["prefabPath"]?.ToObject<string>();

            if (string.IsNullOrEmpty(prefabPath))
            {
                return McpUnitySocketHandler.CreateErrorResponse(
                    "Required parameter 'prefabPath' not provided",
                    "validation_error"
                );
            }

            prefabPath = prefabPath.Trim().Replace('\\', '/');

            if (string.IsNullOrEmpty(AssetDatabase.AssetPathToGUID(prefabPath)))
            {
                return McpUnitySocketHandler.CreateErrorResponse(
                    $"Prefab asset not found at path '{prefabPath}'",
                    "not_found_error"
                );
            }

            GameObject prefab = AssetDatabase.LoadAssetAtPath<GameObject>(prefabPath);
            if (prefab == null)
            {
                return McpUnitySocketHandler.CreateErrorResponse(
                    $"Asset at path '{prefabPath}' is not a prefab GameObject",
                    "validation_error"
                );
            }

            AssetDatabase.OpenAsset(prefab);

            return new JObject
            {
                ["success"] = true,
                ["type"] = "text",
                ["message"] = $"Opened prefab asset at '{prefabPath}'",
                ["prefabPath"] = prefabPath
            };
        }
    }
}

File: Packages/mcp-unity/Server~/src/tools/prefabOpenTool.ts
import { McpServer } from '@modelcontextprotocol/sdk/server/mcp.js';
import { McpUnity } from '../unity/mcpUnity.js';
import { McpUnityError, ErrorType } from '../utils/errors.js';
import * as z from 'zod';
import { Logger } from '../utils/logger.js';
import { CallToolResult } from '@modelcontextprotocol/sdk/types.js';

const toolName = 'prefab_open';
const toolDescription = 'Opens a prefab asset in Prefab Mode.';

const paramsSchema = z.object({
  prefabPath: z.string().describe('The path of the prefab asset to open')
});

export function registerPrefabOpenTool(server: McpServer, mcpUnity: McpUnity, logger: Logger) {
  logger.info(`Registering tool: ${toolName}`);

  server.tool(
    toolName,
    toolDescription,
    paramsSchema.shape,
    async (params: z.infer<typeof paramsSchema>) => {
      try {
        logger.info(`Executing tool: ${toolName}`, params);
        const result = await toolHandler(mcpUnity, params);
        logger.info(`Tool execution successful: ${toolName}`);
        return result;
      } catch (error) {
        logger.error(`Tool execution failed: ${toolName}`, error);
        throw error;
      }
    }
  );
}

async function toolHandler(mcpUnity: McpUnity, params: z.infer<typeof paramsSchema>): Promise<CallToolResult> {
  if (!params.prefabPath || params.prefabPath.trim() === '') {
    throw new McpUnityError(
      ErrorType.VALIDATION,
      "Required parameter 'prefabPath' not provided"
    );
  }

  const response = await mcpUnity.sendRequest({
    method: toolName,
    params: { prefabPath: params.prefabPath }
  });

  if (!response.success) {
    throw new McpUnityError(
      ErrorType.TOOL_EXECUTION,
      response.message || 'Failed to open prefab'
    );
  }

  return {
    content: [{
      type: response.type || 'text',
      text: response.message || 'Successfully opened prefab'
    }],
    data: {
      prefabPath: response.prefabPath
    }
  };
}
